File: BackEnd/src/domain/usecases/community/GetPostCommentsUseCase.ts
/**
 * @file GetPostCommentsUseCase.ts
 * @description Caso de uso para obtener los comentarios activos de un post.
 * Enriquece cada comentario con los datos del autor (N+1 — aceptable para TFG).
 * @module Community
 * @layer Domain
 *
 * @implements {IGetPostCommentsUseCase}
 * @injectable
 * @dependencies ICommentRepository, IUserRepository, ICommentMapper
 */

import { injectable, inject } from 'inversify';
import { TYPES } from '../../../core/types.js';
import type { IGetPostCommentsUseCase } from '../../interfaces/usecases/community/IGetPostCommentsUseCase.js';
import type { ICommentRepository } from '../../repositories/ICommentRepository.js';
import type { IUserRepository } from '../../repositories/IUserRepository.js';
import type { ICommentMapper } from '../../../data/IMappers/ICommentMapper.js';
import type { CommentResponseDTO } from '../../dtos/community/comment-response.dto.js';

/**
 * Obtiene los comentarios activos de un post, enriquecidos con datos del autor.
 *
 * @implements {IGetPostCommentsUseCase}
 * @injectable
 * @dependencies ICommentRepository, IUserRepository, ICommentMapper
 */
@injectable()
export class GetPostCommentsUseCase implements IGetPostCommentsUseCase {
  constructor(
    @inject(TYPES.ICommentRepository) private readonly commentRepo: ICommentRepository,
    @inject(TYPES.IUserRepository)    private readonly userRepo: IUserRepository,
    @inject(TYPES.ICommentMapper)     private readonly mapper: ICommentMapper,
  ) {}

  /**
   * @param postId — Id del post cuyos comentarios se solicitan.
   * @returns Lista de CommentResponseDTOs ordenada cronológicamente.
   */
  async execute(postId: string): Promise<CommentResponseDTO[]> {
    const comments = await this.commentRepo.findByPostId(postId);

    // TFG: N+1 aceptable para datasets pequeños. En producción usar $lookup.
    const result: CommentResponseDTO[] = [];
    for (const comment of comments) {
      const author = await this.userRepo.findById(comment.userId);
      result.push(this.mapper.toResponseDTO(
        comment,
        author?.name ?? 'Usuario eliminado',
        author?.photo,
      ));
    }

    return result;
  }
}

File: FrontEnd/lib/domain/usecases/community/get_post_comments_use_case.dart
/// @file get_post_comments_use_case.dart
/// @description Implementación del caso de uso para obtener los comentarios de un post.
/// @module Community
/// @layer Domain
library;

import '../../entities/comment.dart';
import '../../interfaces/usecases/community/i_get_post_comments_use_case.dart';
import '../../repositories/i_post_repository.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// GET POST COMMENTS USE CASE
// ═══════════════════════════════════════════════════════════════════════════════

/// Obtiene los comentarios de un post ordenados por fecha de creación.
///
/// [implements] IGetPostCommentsUseCase
/// [injectable] registrar en container.dart como singleton.
/// [dependencies] IPostRepository.
class GetPostCommentsUseCase implements IGetPostCommentsUseCase {
  final IPostRepository _repository;

  const GetPostCommentsUseCase({required IPostRepository repository})
      : _repository = repository;

  @override
  Future<List<Comment>> execute(String postId) {
    return _repository.getComments(postId);
  }
}
